### DISCUSSION TITLE
File / socket handle counts vs just a "general" handle count

### DISCUSSION THREAD

**Discussion by lukebakken** [rabbitmq/fshc#6] [General]
I don't know if @the-mikedavis used this as a guide for his implementation of file handle counts in this project ...

https://github.com/rabbitmq/rabbitmq-server/issues/8700#issuecomment-1652409884

...but as he pointed [out in his comments for the feature](https://github.com/rabbitmq/fshc/commit/c4108cd8aeea863a0953a4be1401e8d1dcc90744), it uses an unofficial NT API as well as requires enumerating all of the handles in the system.

Since file/socket handle counts don't really matter as much on Windows as other systems, what do you think about having an option to use `GetProcessHandleCount` to get a general idea of handles for a PID?

cc @the-mikedavis @michaelklishin

**Comment by the-mikedavis** [comment]
I tried out GetProcessHandleCount here: https://github.com/rabbitmq/fshc/commit/4b2ae78c2525183bd80e1952b4c0038db4c71511

Feel free to experiment with that patch and check out the handle counts. From what I've seen so far it looks like the total count is usually much higher than the file handle count alone, mostly with Event handles according to handle.exe.

I see that branch taking 25ms on a release build vs the ~100ms it takes on master (vs ~225ms for handle.exe).

**Comment by michaelklishin** [comment]
So, an option to return the total count? Sounds good to me.

**Comment by lukebakken** [comment]
Thanks @the-mikedavis! I'm going to test that out soon. I'm hoping the non-file, non-socket handle count will stay somewhat stable as various other RabbitMQ bits increase in count (vhosts, connections, etc) and that `GetProcessHandleCount` can be a "good enough" option. But, I think we should be able to choose between the two with an option if that's not too much trouble.

Another reason to have two options is the "Missing `File:`" error we see with `handle.exe`. I have no idea why, but sometimes `handle.exe` does not return all of the expected counts and RabbitMQ logs the error. I have seen this on my local machine as well, and asked @markrussinovich about it but didn't get a response. Since we don't have the code I'm assuming that it _might_ be something with `NtQuerySystemInformation`.

**Comment by michaelklishin** [comment]
@lukebakken can you please elaborate on how RabbitMQ would use this option? We report file and socket counts as node metrics.

**Reply by lukebakken** [reply]
> We report file and socket counts as node metrics

Right, but we don't _do anything_ with that data so it's not as important as something like disk space. Windows also doesn't have the sort of limits that POSIX has.

It may confuse users to see the same number for both files and sockets.

Also see [this comment](https://github.com/rabbitmq/fshc/discussions/6#discussioncomment-6791696) for why we may want two methods for getting handle counts.

**Reply by michaelklishin** [reply]
OK, no objections to having both options

**Comment by michaelklishin** [comment]
#7
